File: 백준/Bronze/2605. 줄 세우기/줄 세우기.py
import sys
data_temp = sys.stdin if sys.platform == 'linux' else open('입력.txt', 'r')
input_data = data_temp.read().splitlines()

def solution (data) :
  [N], arr = map(lambda x : list(map(int,x.split())), data)
  
  result = [1]
  for i in range(1, N) :
    n = arr[i]
    if n == 0 : 
      result.append(i + 1)
    else : 
      _len = len(result) - n
      result = [*result[:_len], i + 1, *result[_len:]]
    
  print(' '.join(map(str,result)))
  
solution(input_data)

File: 백준/Gold/2252. 줄 세우기/줄 세우기.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "입력.txt")
  .toString()
  .trim()
  .split("\n");

function solution(data) {
  const [[N, _], ...arr] = data.map((el) => el.split(" ").map(Number));

  const graph = Array.from({ length: N + 1 }, () => []);
  const rank = Array.from({ length: N + 1 }, () => 0);

  arr.forEach(([a, b]) => {
    graph[a].push(b);
    rank[b]++;
  });

  const queue = [];
  let queueIdx = 0;

  for (let i = 1; i < N + 1; i++) {
    if (!rank[i]) queue.push(i);
  }

  const result = [];
  while (queueIdx < queue.length) {
    const node = queue[queueIdx++];
    result.push(node);
    graph[node].forEach((next) => {
      rank[next]--;
      if (!rank[next]) queue.push(next);
    });
  }

  console.log(...result);
}

solution(input);
